rasa/nlu/classifiers/diet_classifier.py

```diff
@@ -616,90 +630,103 @@ def _create_model_data(
+
+        attributes_to_consider = [TEXT]
+        if training and self.component_config[INTENT_CLASSIFICATION]:
+            # we don't have any intent labels during prediction, just add them during
+            # training
+            attributes_to_consider.append(label_attribute)
+        if training and self.component_config[ENTITY_RECOGNITION]:
+            # we don't have any entity tags during prediction, just add them during
+            # training
+            attributes_to_consider.append(ENTITIES)
+
+        if training and label_attribute is not None:
+            # only use those training examples that have the label_attribute set
+            # during training
+            training_data = [
+                example for example in training_data if label_attribute in example.data
+            ]
 
-        # TODO: simplify model data creation
-        #   convert training data into a list of attribute to features and reuse some
-        #   of the methods of TED (they most likely need to change a bit)
+        if not training_data:
+            # no training data are present to train
+            return RasaModelData()
 
-        features = defaultdict(lambda: defaultdict(list))
-        label_ids = []
+        features_for_examples = model_data_utils.convert_training_examples(
+            training_data,
+            attributes_to_consider,
+            entity_tag_specs=self._entity_tag_specs,
+            featurizers=self.component_config[FEATURIZERS],
+            bilou_tagging=self.component_config[BILOU_FLAG],
+        )
+        attribute_data, _ = model_data_utils.convert_to_data_format(
+            features_for_examples, consider_dialogue_dimension=False
+        )
 
-        for example in training_data:
-            if label_attribute is None or example.get(label_attribute):
-                text_features = self._extract_features(example, TEXT)
-                for feature_key, feature_value in text_features.items():
-                    features[TEXT][feature_key].append(feature_value)
+        model_data = RasaModelData(
+            label_key=self.label_key, label_sub_key=self.label_sub_key
+        )
+        model_data.add_data(attribute_data)
+        model_data.add_lengths(TEXT, SEQUENCE_LENGTH, TEXT, SEQUENCE)
 
-            # only add features for intent labels during training
-            if training and example.get(label_attribute):
-                label_features = self._extract_features(example, label_attribute)
-                for feature_key, feature_value in label_features.items():
-                    features[LABEL][feature_key].append(feature_value)
+        self._add_label_features(
+            model_data, training_data, label_attribute, label_id_dict, training
+        )
 
-                if label_id_dict:
-                    label_ids.append(label_id_dict[example.get(label_attribute)])
+        # make sure all keys are in the same order during training and prediction
```

Comment:
I would add in the comment why they need to sorted?
I believe it's because when we pass it to the model and construct tf tensors out of them we rely on the indices to slice the incoming batch data and reconstruct features for all keys and sub-keys of RasaModelData?

Reply:
@tabergma could you please take this over

Reply:
As RasaModelData is now a dict of dict the order especially for the sub keys is not guaranteed to the be the same during construction. It is fine during training but we need to make sure that the order is the same during prediction as we rely on the indices when converting RasaModelData to the actual batch.